storage/src/mmr/iterator.rs

```diff
@@ -49,33 +49,32 @@ impl PeakIterator {
     ///
     /// The implementation verifies that peaks in the MMR of the given size have strictly decreasing
     /// height, which is a necessary condition for MMR validity.
-    pub(crate) fn check_validity(size: u64) -> bool {
-        if size == 0 {
+    pub(crate) fn check_validity(mut size: u64) -> bool {
+        if size <= 1 {
             return true;
         }
-        let start = u64::MAX >> size.leading_zeros();
-        let mut two_h = 1 << start.trailing_ones();
-        let mut node_pos = start - 1;
-        while two_h > 1 {
-            if node_pos < size {
-                if two_h == 2 {
-                    // If this peak is a leaf yet there are more nodes remaining, then this MMR is
-                    // invalid.
-                    return node_pos == size - 1;
-                }
-                // move to the right sibling
-                node_pos += two_h - 1;
-                if node_pos < size {
-                    // If the right sibling is in the MMR, then it is invalid.
-                    return false;
-                }
-                continue;
+
+        // Consider each mountain from left to right.
+        // For each, subtract its size from `size`.
+        // We should find that the size eventually reaches 0
+        // (the last mountain has more than 1 element) or 1
+        // (the last mountain has 1 element).
+        // If the size never reaches 0, there are "left over"
+        // nodes and there isn't an MMR with the given `size`.
+
+        // Height of the root of the smallest perfect binary
+        // tree containing `size` (leaf is height 0)
+        let mut height = 63 - size.leading_zeros();
```

Comment:
While this is more clear in isolation, given the rest of the iterator implementations maintain 2^(height+1) as "two_h" instead of converting height directly with each iteration, it might be best to stick with that & save a few cycles of computation each iteration.